### PR TITLE
Cell map

### PR DESCRIPTION
**Pull request by valeriaRaffuzzi**
Cell map for tallies. 

It maps the cells with their uniqueID, retrieved from cellShelf.

One issue with this map is that it requires a pointer to geometryStd, and the only way I could think of was through geometryReg, but this needs a geomIdx as an input (or the name of the geometry). So right now this works under the assumption that there is only one geometry (geomIdx is hardcoded and equals 1), and that the type is geometryStd. This is well specified in the documentation. It is not a problem now, because we can only read one geometry, but if this changes in the future we will have to think about it.

### CODE REVIEW

File: Tallies/TallyMaps/cellMap_class.f90 [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    call self % binMap % init(N)
+    allocate(self % cellIndices(N))
+
+    ! Get pointer to geometry
+    ! ASSUMES THERE IS ONLY ONE GEOMETRY, AND THAT IT IS STD
+    geom => geomPtr(1)
+    geomStd => geometryStd_CptrCast(geom)
+
+    ! Load cell indices and bins
+    do i = 1,N
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/cellMap_class.f90:107]:
```suggestion
    do i = 1, N
```

File: Tallies/TallyMaps/cellMap_class.f90 [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    select case (undefined)
+      case('yes','y','true','TRUE','T')
+        trackUndefined = .true.
+
+      case('no', 'n', 'false', 'FALSE', 'F')
+        trackUndefined = .false.
+
+      case default
+        call fatalError(Here, undefined//' is an unrecognised entry!')
+    end select
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/cellMap_class.f90:154]: We really need to make a function for that conversion. Not in this PR but I guess we should start an issue.

File: Tallies/TallyMaps/cellMap_class.f90 [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+  subroutine print(self,out)
+    class(cellMap), intent(in)       :: self
+    class(outputFile), intent(inout) :: out
+    character(nameLen)               :: name
+    integer(shortInt)                :: i
+
+    ! Name the array
+    name = trim(self % getAxisName()) //'Bins'
+
+    call out % startArray(name,[1,self % Nbins])
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/cellMap_class.f90:220]:
```suggestion
    call out % startArray(name, [1, self % Nbins])
```

File: Tallies/TallyMaps/cellMap_class.f90 [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+    ! Print cell indexes
+    do i=1,size(self % cellIndices)
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/cellMap_class.f90:223]:
```suggestion
    do i = 1, size(self % cellIndices)
```

File: Tallies/TallyMaps/cellMap_class.f90 [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+  !!
+  elemental subroutine kill(self)
+    class(cellMap), intent(inout) :: self
+
+    call kill_super(self)
+
+    call self % binMap % kill()
+    self % default = 0
+    self % Nbins = 0
+
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/cellMap_class.f90:268]:
```suggestion
    if (allocated(self % cellIndices) deallocate(self % cellIndices) 
```

File: Tallies/TallyMaps/cellMap_class.f90 [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+  !! See tallyMap for specification
+  !!
+  subroutine print(self,out)
+    class(cellMap), intent(in)       :: self
+    class(outputFile), intent(inout) :: out
+    character(nameLen)               :: name
+    integer(shortInt)                :: i
+
+    ! Name the array
+    name = trim(self % getAxisName()) //'Bins'
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/cellMap_class.f90:218]:
```suggestion
    name = trim(self % getAxisName()) // 'Bins'
```

File: Tallies/TallyMaps/cellMap_class.f90 [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+  !! Constructor
+  !!
+  interface cellMap
+    module procedure cellMap_fromDict
+  end interface
+
+  !!
+  !! Map that divides based on the cell a particle is in
+  !! NOTE: the cell ID must be converted from the IDs in the input file
+  !!       to the uniqueIDs.
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/cellMap_class.f90:28]: Name `uniqueId` is used for the unique number given to each instance of any cell in the geometry. 
For position of the `cell_inter` in the shelf I have been using `cellIdx` 

Basically the number in input file is `cellId`, position in a shelf `cellIdx` and a unique identifier dependent on a position of the isntance in the geometry graph `uniqueId`.

File: Tallies/TallyMaps/Tests/cellMap_test.f90 [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+  !!
+@Test
+  subroutine testMappingNoUndefined(this)
+    class(test_cellMap), intent(inout)       :: this
+    type(particleState)                      :: state
+    integer(shortInt)                        :: i
+    integer(shortInt),dimension(5)           :: bins
+    integer(shortInt),dimension(5),parameter :: EXPECTED_BINS = [1, 0, 2, 3, 0]
+
+    do i = 1,5
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/Tests/cellMap_test.f90:124]:
```suggestion
    do i = 1, size(EXPECTED_BINS)
```
Sorry, C++ induced paranoia ;-)

File: Tallies/TallyMaps/Tests/cellMap_test.f90 [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+  !!
+@Test
+  subroutine testMappingUndefined(this)
+    class(test_cellMap), intent(inout)       :: this
+    type(particleState)                      :: state
+    integer(shortInt)                        :: i
+    integer(shortInt),dimension(5)           :: bins
+    integer(shortInt),dimension(5),parameter :: EXPECTED_BINS = [1, 4, 2, 3, 4]
+
+    do i = 1,5
```
Comment (Mikolaj-A-Kowalski) [inline Tallies/TallyMaps/Tests/cellMap_test.f90:145]:
```suggestion
    do i = 1, size(EXPECTED_BINS)
```

**Review by Mikolaj-A-Kowalski — APPROVED** [review]
Thank you for the map! 
Please pay no attention to all the suggestions regarding spaces :-) Feel completely free to ignore them. For me it just feels nicer with the extra whitespace, but it is completely irrelevant.

**Comment by valeriaRaffuzzi** [comment]
All the changes were accepted. The main ones included checks with fatalErrors in case geometryReg included more than one geometry, and if the geometry stored is not geometryStd. A function was added in geometryReg to return the number of geometries allocated.

**Pull request merged**